prow/crier/controller_test.go

```diff
@@ -0,0 +1,268 @@
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                                 { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                                 { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*prowv1.ProwJob, err error)          { return nil, nil }
+
+func TestController_Run(t *testing.T) {
+	tests := []struct {
+		name         string
+		jobsOnQueue  []string
+		knownJobs    map[string]*prowv1.ProwJob
+		shouldReport func(*prowv1.ProwJob) bool
+
+		expectReports []string
+		expectPatch   bool
+	}{
+		{
+			name:        "reports/patches known job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+			expectReports: []string{"foo"},
+			expectPatch:   true,
+		},
+		{
+			name:        "doesn't report when it shouldn't",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return false
+			},
+		},
+		{
+			name:        "doesn't report nonexistant job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs:   map[string]*prowv1.ProwJob{},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+		//{
+		//	name: "nil job panics",
+		//	jobsOnQueue: []string{"foo"},
+		//	knownJobs: map[string]*prowv1.ProwJob{
+		//		"foo" : nil,
+		//	},
+		//	shouldReport: func(*prowv1.ProwJob) bool {
+		//		return true
+		//	},
+		//},
+		{
+			name:        "doesn't report empty job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+		{
+			name:        "duplicate jobs report once",
+			jobsOnQueue: []string{"foo", "foo", "foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+			expectReports: []string{"foo"},
+			expectPatch:   true,
+		},
+		{
+			name:        "previously-reported job isn't reported",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+						PrevReportStates: map[string]prowv1.ProwJobState{
+							reporterName: prowv1.TriggeredState,
```

Comment:
How about the scenario where reporter X but not Y has reported?

Reply:
Can you explain how reporter X could report but not reporter Y, when a controller is constructed with one reporter?

Do you mean a case where a report call returns multiple, or no, jobs?